docs/docsite/rst/os_guide/windows_performance.rst

```diff
@@ -20,24 +20,50 @@ seconds.
 
 .. code-block:: powershell
 
-    function Optimize-PowershellAssemblies {
-      # NGEN powershell assembly, improves startup time of powershell by 10x
-      $old_path = $env:path
-      try {
-        $env:path = [Runtime.InteropServices.RuntimeEnvironment]::GetRuntimeDirectory()
-        [AppDomain]::CurrentDomain.GetAssemblies() | % {
-          if (! $_.location) {continue}
-          $Name = Split-Path $_.location -leaf
-          if ($Name.startswith("Microsoft.PowerShell.")) {
-            Write-Progress -Activity "Native Image Installation" -Status "$name"
-            ngen install $_.location | % {"`t$_"}
-          }
-        }
-      } finally {
-        $env:path = $old_path
-      }
+    function Optimize-Assemblies {
+        param (
+            [string]$assemblyFilter = "Microsoft.PowerShell.",
+            [string]$activity = "Native Image Installation"
+        )
+        
+        # Save the current $env:path
+        $old_path = $env:path
+        
+        try {
+            # Set $env:path to the runtime directory
+            $env:path = [Runtime.InteropServices.RuntimeEnvironment]::GetRuntimeDirectory()
+            
+            # Get a list of loaded assemblies
+            $assemblies = [AppDomain]::CurrentDomain.GetAssemblies()
+            
+            # Filter assemblies based on the provided filter
+            $filteredAssemblies = $assemblies | Where-Object { $_.Location -match $assemblyFilter }
```

Comment:
```suggestion
            $filteredAssemblies = $assemblies | Where-Object { $_.Location.StartsWith($AssemblyFilter) }
```
By switching to `-match` you've changed this to a RegEx match, and that changes the meaning of the filter, which is important because it's now paramertized in the script, and because the filer is not anchored.

If you want to support RegEx, you'll want to change the default value of the param appropriately, and also if anyone changes that value they need to know how to write it properly as a pattern.

I also think the original loop with the conditional inside was fine; separately iterating over the loop twice is something we might want to avoid.

Alternatively, pipe to `Where-Object` like this and then pipe directly to `ForEach-Object` to then process the items that make it through.

Reply:
We use match because it isn't case sensitive. StartsWith, on the other hand, is.
However to get similar functionality to StartsWith, with anchoring, but without case-sensitivity I can use ilike.
```
        $filteredAssemblies = $assemblies | Where-Object { $_.FullName -ilike "$assemblyFilter*" }

```

Reply:
As far as the piping with where and foreach like you've mentioned, I can agree to that. However I do like a certain level of clarity in the script. 

The optimization of piping it directly into the foreach, saves milliseconds. I don't think it's worth it to remove the error statement to get this benefit. But I'll leave that up to you.

Reply:
You can pass in an argument to StartsWith to make it case insensitive.

```powershell
'Foo'.StartsWith('f')  # False
'Foo'.StartsWith('f', 'OrdinalIgnoreCase')  # True
```